modules/katamari/src/main/ts/ephox/katamari/api/Optional.ts

```diff
@@ -1,141 +1,290 @@
+   */
+  public fold<U>(onNone: () => U, onSome: (value: T) => U): U {
+    if (this.tag) {
+      return onSome(this.value as T);
+    } else {
+      return onNone();
+    }
+  }
 
-  readonly getOrNull: () => T | null;
-  readonly getOrUndefined: () => T | undefined;
   /**
-  - if some: return self
-  - if none: return opt
-  */
-  readonly or: <T2 = T>(opt: Optional<T2>) => Optional<T | T2>;
+   * Determine if this `Optional` object contains a value.
+   */
+  public isSome(): boolean {
+    return this.tag;
+  }
 
-  /** Same as "or", but uses a thunk instead of a value */
-  readonly orThunk: <T2 = T>(makeOption: () => Optional<T2>) => Optional<T | T2>;
+  /**
+   * Determine if this `Optional` object **does not** contain a value.
+   */
+  public isNone(): boolean {
+    return !this.tag;
+  }
 
-  /** Run a function over the 'some' value.
-   *  "map" operation on the Optional functor.
+  // --- Functor (name stolen from Haskell / maths) ---
+
+  /**
+   * Perform a transform on an `Optional` object, **if** there is a value. If
+   * you provide a function to turn a T into a U, this is the function you use
+   * to turn an `Optional<T>` into an `Optional<U>`. If this **does** contain
+   * a value then the output will also contain a value (that value being the
+   * output of `mapper(this.value)`), and if this **does not** contain a value
+   * then neither will the output.
    */
-  readonly map: <U> (mapper: (x: T) => U) => Optional<U>;
+  public map<U>(mapper: (value: T) => U): Optional<U> {
+    if (this.tag) {
+      return Optional.some(mapper(this.value as T));
+    } else {
+      return Optional.none();
+    }
+  }
 
-  /** Run a side effect over the 'some' value */
-  readonly each: (worker: (x: T) => void) => void;
+  // --- Monad (name stolen from Haskell / maths) ---
 
-  /** "bind"/"flatMap" operation on the Optional Bind/Monad.
-   *  Equivalent to >>= in Haskell/PureScript; flatMap in Scala.
+  /**
+   * Perform a transform on an `Optional` object, **if** there is a value.
+   * Unlike the map function earlier in the piece, here the transform itself
+   * also returns an `Optional`.
    */
-  readonly bind: <U> (f: (x: T) => Optional<U>) => Optional<U>;
+  public bind<U>(binder: (value: T) => Optional<U>): Optional<U> {
+    if (this.tag) {
+      return binder(this.value as T);
+    } else {
+      return Optional.none();
+    }
+  }
 
-  /** Does this Optional contain a value that predicate? */
-  readonly exists: (f: (x: T) => boolean) => boolean;
+  // --- Traversable (name stolen from Haskell / maths) ---
 
-  /** Do all values contained in this option match this predicate? */
-  readonly forall: (f: (x: T) => boolean) => boolean;
+  /**
+   * For a given predicate, this function finds out if there **exists** a value
+   * inside this `Optional` object that meets the predicate. In practice, this
+   * means that for empty `Optional`s it returns false (as no predicate-
+   * meeting value exists).
+   */
```

Comment:
Maybe instead of saying `empty` Optionals, as that's not very clear (e.g. you could have an empty array), say something like this?
```suggestion
  /**
   * For a given predicate, this function finds out if there **exists** a value
   * inside this `Optional` object that meets the predicate. In practice, this
   * means that for a `None` it returns false (as no predicate-meeting value
   * exists).
   */
```

Reply:
I've clarified this to "Optionals that don't contain a value" to match up with the other docs